File: java/src/test/java/it/giannibombelli/wsc2026/booking/application/usecases/BookingPlacingTest.java
package it.giannibombelli.wsc2026.booking.application.usecases;

import it.giannibombelli.wsc2026.booking.application.commands.PlaceBooking;
import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;
import it.giannibombelli.wsc2026.booking.infrastructure.SqliteBookingRepository;
import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class BookingPlacingTest {
    private BookingRepository repository;
    private BookingPlacing placement;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("booking");
        repository = new SqliteBookingRepository(dataSource);
        placement = new BookingPlacing(repository, event -> {
        });
    }

    @Nested
    class Construction {
        @Test
        void rejectNullRepository() {
            assertThatThrownBy(() -> new BookingPlacing(null, event -> {
            }))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class Execution {
        @Test
        void rejectNullCommand() {
            assertThatThrownBy(() -> placement.invoke(null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldPersist() {
            BookingId bookingId = EntityId.generate(BookingId::new);
            Money amount = new Money(new BigDecimal("123.45"));
            Description description = new Description("Some description");
            GiftCardId giftCardId = EntityId.generate(GiftCardId::new);

            placement.invoke(new PlaceBooking(bookingId, amount, description, giftCardId));

            Optional<Booking> loaded = repository.findById(bookingId);
            assertThat(loaded).isPresent();
            assertThat(loaded.get().id()).isEqualTo(bookingId);
            assertThat(loaded.get().description()).isEqualTo(description);
            assertThat(loaded.get().giftCardId()).isEqualTo(giftCardId);
        }
    }
}

File: typescript/tests/booking/application/usecases/bookingPlacing.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Description } from '@/common/domain/primitive/description.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { placeBooking, type PlaceBooking } from '@/booking/application/commands/placeBooking.js';
import { BookingPlacing } from '@/booking/application/usecases/bookingPlacing.js';
import { BookingId } from '@/booking/domain/booking/bookingId.js';
import { SqliteBookingRepository } from '@/booking/infrastructure/sqliteBookingRepository.js';
import { GiftCardReference } from '@/booking/domain/primitive/giftCardReference.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';
import type { BookingEvent } from '@/booking/domain/events/bookingEvent.js';
import type { EventPublisher } from '@/common/application/events/eventPublisher.js';

describe('BookingPlacing', () => {
  let repository: SqliteBookingRepository;
  let placement: BookingPlacing;

  const noOpPublisher: EventPublisher<BookingEvent> = {
    publish: () => {},
  };

  beforeEach(() => {
    const database = DatabaseSetup.initializeInMemoryDb('booking');
    repository = new SqliteBookingRepository(database);
    placement = new BookingPlacing(repository, noOpPublisher);
  });

  describe('construction', () => {
    it('reject null repository', () => {
      expect(() => new BookingPlacing(null as unknown as SqliteBookingRepository, noOpPublisher)).toThrow();
    });
  });

  describe('execution', () => {
    it('reject null command', () => {
      expect(() => placement.invoke(null as unknown as PlaceBooking)).toThrow();
    });

    it('should persist', () => {
      const bookingId = generateId((value) => new BookingId(value));
      const amount = new Money(123.45);
      const description = new Description('Some description');
      const giftCardReference = new GiftCardReference(Uuid.generate());

      placement.invoke(placeBooking(bookingId, amount, description, giftCardReference));

      const loaded = repository.findById(bookingId);
      expect(loaded).not.toBeNull();
      expect(loaded!.id()).toEqual(bookingId);
      expect(loaded!.description()).toEqual(description);
      expect(loaded!.giftCardReference()).toEqual(giftCardReference);
    });
  });
});
